### PR TITLE
Ignore WithTenantId for B2C, Refactor AADAuthority Tenant Checks

### LINKED ISSUE #3471: [Bug] ResolveAuthority in AbstractApplicationBuilder doesn't take into account B2C authorities
**Which version of MSAL.NET are you using?**
Microsoft.Identity.Web 1.25.1 with MSAL.NET 4.42.0?

**Actual behavior**
See https://github.com/AzureAD/microsoft-identity-web/issues/1805. MSAL throws AuthorityTenantSpecifiedTwice exception when domain and tenant ID are specified. [ResolveAuthority](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/4eba04cec6fbc68fbe4b10e5b7b3f6fcc2eece25/src/client/Microsoft.Identity.Client/AppConfig/AbstractApplicationBuilder.cs#L563-L575) doesn't take into account B2C authorities.

**Expected behavior**
Maybe disregard tenant ID if authority is B2C or at least a better error message.

**Possible solution**
Maybe refactor [ResolveAuthority](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/4eba04cec6fbc68fbe4b10e5b7b3f6fcc2eece25/src/client/Microsoft.Identity.Client/AppConfig/AbstractApplicationBuilder.cs#L563-L575) with a B2CAuthority check. Or make the error message clearer.

### PR DESCRIPTION
**Pull request by MichelZ**
Fixes #3471

- Ignore WithTenantId on B2C Authority
- Refactor AAD Tenant "Type" detections

**Testing**
Unit Test added testing successful Build of PCA WithTenantId

### CODE REVIEW

File: tests/Microsoft.Identity.Test.Unit/PublicApiTests/PublicClientApplicationTestsWithB2C.cs [hunk 1]
```diff
@@ -131,6 +131,23 @@ public async Task B2CAcquireTokenWithValidateAuthorityTrueAndRandomAuthorityTest
             }
         }
 
+        [TestMethod]
+        [TestCategory("B2C")]
+        public void B2CIgnoreWithTenantId_Success()
+        {
+            using (var harness = base.CreateTestHarness())
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit/PublicApiTests/PublicClientApplicationTestsWithB2C.cs:138]: No need to mock anything to validate the authority.

File: tests/Microsoft.Identity.Test.Unit/PublicApiTests/PublicClientApplicationTestsWithB2C.cs [hunk 2]
```diff
@@ -131,6 +131,23 @@ public async Task B2CAcquireTokenWithValidateAuthorityTrueAndRandomAuthorityTest
...
+            {
+                PublicClientApplication app = PublicClientApplicationBuilder.Create(TestConstants.ClientId)
+                    .WithAuthority(new Uri(TestConstants.B2CCustomDomain), true)
+                    .WithHttpManager(harness.HttpManager)
+                    .WithCachePartitioningAsserts(harness.ServiceBundle.PlatformProxy)
+                    .WithTenantId(Guid.NewGuid().ToString("D"))
+                    .BuildConcrete();
+                
+                Assert.IsTrue(true);
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit/PublicApiTests/PublicClientApplicationTestsWithB2C.cs]: You can assert the value `app.Authority`

File: tests/Microsoft.Identity.Test.Unit/PublicApiTests/PublicClientApplicationTestsWithB2C.cs [hunk 3]
```diff
@@ -131,6 +131,23 @@ public async Task B2CAcquireTokenWithValidateAuthorityTrueAndRandomAuthorityTest
...
+        [TestMethod]
+        [TestCategory("B2C")]
+        public void B2CIgnoreWithTenantId_Success()
+        {
+            using (var harness = base.CreateTestHarness())
+            {
+                PublicClientApplication app = PublicClientApplicationBuilder.Create(TestConstants.ClientId)
+                    .WithAuthority(new Uri(TestConstants.B2CCustomDomain), true)
+                    .WithHttpManager(harness.HttpManager)
+                    .WithCachePartitioningAsserts(harness.ServiceBundle.PlatformProxy)
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit/PublicApiTests/PublicClientApplicationTestsWithB2C.cs]: No need for this, there won't be any caching operations.

File: tests/Microsoft.Identity.Test.Unit/PublicApiTests/PublicClientApplicationTestsWithB2C.cs [hunk 4]
```diff
@@ -131,6 +131,23 @@ public async Task B2CAcquireTokenWithValidateAuthorityTrueAndRandomAuthorityTest
...
         }
 
+        [TestMethod]
+        [TestCategory("B2C")]
+        public void B2CIgnoreWithTenantId_Success()
+        {
+            using (var harness = base.CreateTestHarness())
+            {
+                PublicClientApplication app = PublicClientApplicationBuilder.Create(TestConstants.ClientId)
+                    .WithAuthority(new Uri(TestConstants.B2CCustomDomain), true)
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit/PublicApiTests/PublicClientApplicationTestsWithB2C.cs]: B2C have custom domain authorities like this one but also `TestConstants.B2CAuthority`

File: tests/Microsoft.Identity.Test.Unit/PublicApiTests/PublicClientApplicationTestsWithB2C.cs [hunk 5]
```diff
@@ -131,6 +131,18 @@ public async Task B2CAcquireTokenWithValidateAuthorityTrueAndRandomAuthorityTest
             }
         }
 
+        [TestMethod]
+        [TestCategory("B2C")]
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit/PublicApiTests/PublicClientApplicationTestsWithB2C.cs:135]: Please add a comment mentioning that this is a test against the bug https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/3471

**Review by bgavrilMS — APPROVED** [review]

File: src/client/Microsoft.Identity.Client/Instance/AadAuthority.cs [hunk 6]
```diff
@@ -65,7 +69,6 @@ internal bool IsCommonOrOrganizationsTenant()
         internal static bool IsCommonOrOrganizationsTenant(string tenantId)
         {
             return !string.IsNullOrEmpty(tenantId) && 
-                tenantId != Constants.ConsumerTenant &&
```
Comment (neha-bhargava) [inline src/client/Microsoft.Identity.Client/Instance/AadAuthority.cs:68]: 👍

**Review by neha-bhargava — APPROVED** [review]
Looks good

**Pull request merged**